### ISSUE TITLE
Add support for VIRTUAL and STORED client side computed columns

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#9879]
### Purpose

Many dialects have native support for server side computed columns using syntaxes such as `GENERATED ALWAYS AS`. These are useful for:

- Replacing triggers to produce identity values / sequence generated values or UUID primary key values
- Calculate some expression that can be used in an index. It's generally a cleaner approach than function based indexes, as the use site does not have to remember and repeat the expression in *every* query in order for the computed column's index to apply.

If a dialect doesn't support these features, or in some cases where the feature is not good enough for various reasons (e.g. Oracle doesn't support materialised computed columns, which would be useful if the computation is CPU intensive), it would be nice to have this feature in jOOQ, client side.

### Comparison with `RecordListener`

Users can already implement it using `RecordListener`, but this feature would affect also any explicitly constructed `INSERT`, `UPDATE` or `MERGE` statement, including parsed ones, if meta data lookup is applied.

Unlike `RecordListener`, which has access to the entire `Record` being inserted, the computation expression cannot access any *values* from other columns, although accessing *expressions* based on other columns is possible.

### Tasks

The following changes are needed:

- [x] Additional meta information on the `DataType` meta model
  - [x] A new `GenerationLocation` enum with values `CLIENT` and `SERVER` (default)
  - [x] A new `Generator<T> extends Function<GeneratorContext, Field<T>>, Serializable` type
    - [x] A `GeneratorContext <: Scope` argument object
      - [x] Providing access to a `Configuration`
      - [x] Providing access to statement information (`INSERT`, `UPDATE`, `SELECT`)
      - [x] Providing access to the target `Field<T>` expression that is being generated
    - [x] Or better: `Generator<R extends Record, T extends Table<R>, U> extends Function<T, Field<U>>, Serializable`, such that lambdas can reference aliased tables?
      - [x] It doesn't seem to be necessary as a `Generator` can always access the aliased table via `this` (lambda) or `MyTable.this` (anonymous class). On the other hand, that would require a `Generator` to capture the instance, which can cause other issues.
  - [x] A `DataType.generatedAlwaysAs(Generator<T>)` overload, which works only with `GenerationLocation.CLIENT`
  - [x] ~A `DataType.default(Generator<T>)` overload, which works only with `GenerationLocation.CLIENT`~: Postponed: https://github.com/jOOQ/jOOQ/issues/13443
- [x] Code generation support to produce synthetic `GENERATED` expressions
  - [x] A `forcedType` configuration element `generator`, which can reference a class or contain inline code
  - [x] It should be possible to have a forced type only with a `generator`. No `name`, no `userType`
  - [x] `generator` applies to
    - [x] `TableField` (first draft)
    - [x] ~`IN` parameters (maybe in a later release)~: Postponed https://github.com/jOOQ/jOOQ/issues/13444
    - [x] ~`UDT` attributes (maybe in a later release)~: Postponed https://github.com/jOOQ/jOOQ/issues/13444
  - [x] Tests in Java, Scala, Kotlin
- [x] ~Settings to enable / disable the feature at runtime~ (this is quite some work, let's wait for a customer requesting this. The workarounds are simple. Just create a plain SQL column expression)
- [x] Implementation logic
- [x] Tests for `VIRTUAL` semantics:
  - [x] Simple query
    - [x] Projecting virtual columns
    - [x] Projecting expressions based on virtual columns
      - [x] Function calls
      - [x] Binary expressions
      - [x] Aggregate functions
      - [x] Window functions
    - [x] Projecting aliased virtual columns
    - [x] Projecting transitive computations (computed column depends on another computed column)
  - [x] Implicit join
    - [x] Computed implicit join referencing another computed column
  - [x] Multisets
    - [x] Uncorrelated
    - [x] Correlated
    - [x] Nested
    - [x] ~MultisetAgg~ (this doesn't really make much sense as a client side computed virtual column)
  - [x] Table aliases
    - [x] Especially when used with implicit joins
    - [x] These seem to work implicitly if the `Generator` captures a `this` reference of the table, which may be aliased already!
  - [x] `RETURNING` clauses
- [x] Tests for `STORED` semantics:
  - [x] `INSERT`
    - [x] `INSERT .. VALUES`
      - [x] With alternative column order
    - [x] `INSERT .. SELECT`
    - [x] Multi row `INSERT`
    - [x] `INSERT .. ON DUPLICATE KEY ..`
      - [x] `IGNORE`
        - [x] Native implementation
        - [x] Merge emulation
      - [x] `UPDATE`
        - [x] Native implementation (we might be able to leverage a peculiar MySQL "feature" where `SET` clauses aren't evaluated atomically, see https://twitter.com/lukaseder/status/1507019871569125381)
          - [x] Reordering columns to 1) user `SET` clauses, 2) computed columns
        - [x] Merge emulation with `SET ROW = ..`
        - [x] Merge emulation
    - [x] `INSERT .. ON CONFLICT DO ..`
      - [x] `NOTHING`
        - [x] Native implementation
        - [x] Merge emulation
      - [x] `UPDATE`
        - [x] Native implementation (probably depends on https://github.com/jOOQ/jOOQ/issues/6769)
          - [x] `SET ..`
          - [x] `SET ROW = ROW`
          - [x] `SET ROW = (SELECT)`
        - [x] Merge emulation with `SET ROW = ..`
          - [x] `SET ..`
          - [x] `SET ROW = ROW`
          - [x] `SET ROW = (SELECT)`
        - [x] Merge emulation
    - [x] `INSERT .. RETURNING`
  - [x] `UPDATE`
    - [x] `UPDATE .. SET ..`
    - [x] `UPDATE .. SET ROW = ROW`
    - [x] `UPDATE .. SET ROW = SELECT`
    - [x] Computation column inclusion
      - [x] Trivial solution would re-compute all columns
    - [x] `UPDATE .. RETURNING`
  - [x] `UpdatableRecord` operations?
  - [x] Combine with converters / bindings
  - [x] The SQL transformation should have a simple and sophisticated version
    - [x] Sophisticated is everything we've discussed so far, and it can handle arbitrary expressions from the `Generator`
- [x] ~Synthetic `CREATE TABLE` and `ALTER TABLE` DDL support for the interpreter~ (maybe later)

### Related work / follow up work:

- https://github.com/jOOQ/jOOQ/issues/13411
- https://github.com/jOOQ/jOOQ/issues/13418

### Caveats:

- [x] MySQL `UPDATE` doesn't run the `SET` clause atomically, see https://twitter.com/lukaseder/status/1507019364800688134, https://stackoverflow.com/q/37649/521799. This has 2 implications:
  - [x] Computed columns must be calculated lexically after all user-provided `SET` clauses
- [x] It would be *very helpful* if we could guarantee that each `Generator` is invoked exacly once per query rendering, in order to produce more predictable results, e.g. when a `Generator` produces a timestamp.
More caveats and known issues here:

- https://github.com/jOOQ/jOOQ/issues/13685

### Examples:

**Independent computed expressions**

With a hypothetical synthetic table definition like this:

```sql
CREATE TABLE t (
  i INT,
  j INT CLIENT GENERATED ALWAYS AS 1,
  k INT CLIENT GENERATED BY DEFAULT AS 2
);
```

The following statements...

```sql
INSERT INTO t (i, j) VALUES (0, 0);
INSERT INTO t (i, j, k) VALUES (0, 0, 0);
INSERT INTO t (i, j, k) SELECT 0, 0, 0;
UPDATE t SET i = 0, j = 0, k = 0;
```

... would be transformed to these ones

```sql
INSERT INTO t (i, j, k) VALUES (0, 1, 2);
INSERT INTO t (i, j, k) VALUES (0, 1, 0);
INSERT INTO t (i, j, k) SELECT i, 1, k FROM (SELECT 0, 0, 0) t (i, j, k);
UPDATE t SET i = 0, j = 1, k = 0;
```

**Dependent computed expressions**

Just like server side computed columns, it should be possible to form expressions based on other columns, e.g. in the aforementioned hypothetical syntax:

```sql
CREATE TABLE t (
  i INT,
  j INT CLIENT GENERATED ALWAYS AS i + 1,
  k INT CLIENT GENERATED BY DEFAULT AS i + 2
);
```

The following statements...

```sql
INSERT INTO t (i, j) VALUES (0, 0);
INSERT INTO t (i, j, k) VALUES (0, 0, 0);
INSERT INTO t (i, j, k) SELECT 0, 0, 0;
UPDATE t SET i = 0, j = 0, k = 0;
```

... would be transformed to these ones

```sql
INSERT INTO t (i, j, k) SELECT i, i + 1, i + 2 FROM (VALUES (0)) t (i);
INSERT INTO t (i, j, k) SELECT i, i + 1, k FROM (VALUES (0, 0)) t (i, k);
INSERT INTO t (i, j, k) SELECT i, i + 1, k FROM (SELECT 0, 0, 0) t (i, j, k);
UPDATE t 
SET (i, j, k) = (SELECT i, i + 1, k FROM (SELECT 0, 0, 0) t (i, j, k));
```

The `UPDATE .. SET ROW = (SELECT ..)` emulation will be needed for this task, to be investigated: https://github.com/jOOQ/jOOQ/issues/10523 

**More tricky cases**

The `UPDATE .. SET ROW = (SELECT ..)` clause is being used to ensure atomic computation of all client side computed expressions. This means that existing `SET ROW = (SELECT ..)` clauses might be a bit more tricky to handle:

**Other**

There will obviously be edge cases which are not properly handled in the above examples yet.

The following statements...

```sql
UPDATE t SET (i, j, k) = (0, 0, 0);
UPDATE t SET (i, j, k) = (SELECT 0, 0, 0);
UPDATE t SET 
  (i, j) = (SELECT 0, 0),
  k = 0;
```

... would be transformed to these ones

```sql
-- SET ROW = ROW will just first flatten the ROW and then perform the usual emulation
UPDATE t SET (i, j, k) = (SELECT i, i + 1, k FROM (SELECT 0, 0, 0) t (i, j, k));

-- SET ROW = SELECT looks the same as the previous one, if there's only a single assignment
UPDATE t SET (i, j, k) = (SELECT i, i + 1, k FROM (SELECT 0, 0, 0) t (i, j, k));

-- But once we support multiple ROW assignments, some additional trickery will be needed
UPDATE t SET (i, j, k) = (
  SELECT i, i + 1, k 
  FROM 
    (SELECT 0, 0) t (i, j), -- ROW assignment
    (SELECT 0) u (k)        -- k assignment
);
```

See also: https://github.com/jOOQ/jOOQ/issues/13330

### `GenerationOption`

By default, the `GenerationOption.STORED` flag is applied, meaning the value is generated upon `INSERT`, `UPDATE`, `MERGE` directly in relevant SQL statements. However, `GenerationOption.VIRTUAL` is also possible, in case of which the computation takes place in every `SELECT` statement, as soon as the column is included.

- [x] Support `STORED` (when the forced type matches an actual column)
- [x] Support `VIRTUAL` (when the forced type matches a synthetic column). Prerequisites:
  - [x] https://github.com/jOOQ/jOOQ/issues/13434

**Comment by alf** [comment]
As described in my comment on #12722, I’d like something like this functionality. The difference between what’s described here and what I’m interested in, is that I’m not currently at liberty to change the database tables. Thus the client side calculated value is not stored in the database.

**Comment by lukaseder** [comment]
> Thus the client side calculated value is not stored in the database.

That `SYNTHETIC` keyword and the table definition are just hypothetical syntax to describe the feature idea. The actual configuration will be code generation XML, as always (though future jOOQ versions might support such an "annotated" DDL syntax)

**Comment by lukaseder** [comment]
@alf: I'm currently implementing support for this. You can track progress in the updated issue description. So far, it looks as though the feature can be implemented quite well in most dialects. There might be some hard to solve edge cases when computing columns in `MERGE`, to be investigated.

In order to better understand your use-cases, and tailor integration tests around them, do you perhaps have some time to share a few examples of what kinds of computations you'd like to do with this feature?

**Comment by alf** [comment]
That’s fantastic to hear. I’m a bit short on time now, but I can try to add more later.


We have created a function that generates surrogate keys that follows the [Object Identification](https://relay.dev/docs/guides/graphql-server-specification/#object-identification) part of the GraphQL Server Spec from Relay. The function takes a Table and optionally a Key and returns a SelectField<String>.

While this works well, I believe it would be even better if we could expose this as client side generated columns. This would increase discoverability, and we can refer to the resulting columns in documentation and configuration. Depending on how jOOQ solves this, it could also make queries easier and clearer:

```
dsl.select(T.relay_id)
   .from(T)
   .where(T.relay_id.eq(input))
```

as opposed to:

```
dsl.select(getId(T))
   .from(T)
   .where(hasId(T, input))
```

I have other thoughts as well, I’ll try to get back to this later today.

**Comment by alf** [comment]
Expanding on my previous comment and giving some more context.

We’re working on a Student Information System that was conceived around 1993 and has been worked on and improved since. Today it provides critical functionality for almost the entire higher education sector in Norway.

For many years now our customers have had growing needs for integration and automation. We’ve had problems delivering satisfactory on these requirements on previous API building attempts. This time around we’re trying to leverage the database in a better way and have found jOOQ to be a perfect fit. 

As we’ve gotten more familiar with jOOQ and seen what’s possible using rows and multisets, the idea of client side calculated columns is really exciting. The main reason for this is that we have a layer of views between the database and jOOQ. This affords us a way to change the underlying data model without affecting our applications etc. While we could use calculated columns in these views, we’ve found that this forces us to create instead of triggers. This is tedious and error prone. Currently we’re working around this problem by creating read only views with the calculated values we need, and then we use implicit joins to access these values. This is tedious and hard to discover. Calculated columns in Java would solve this in the same way that forcedTypes and embeddable types helps us convert existing columns to more specific Java types. Having this functionality would complete the Trinity and make jOOQ divine 😂

The idea of moving the calculations to Java also resonates well with a lot of our developers who are uncomfortable with such advanced database usage. If they could implement an interface and register this on a table, and this would lead to one or more extra columns being generated, I believe they would be very happy. This would be easy to unit test and develop and maintain. That’s a great boon to us.

I can add some concrete examples of actual logic we have that we would like to move to client generated columns when this is possible, but that must wait until Monday.

**Comment by alf** [comment]
One potential problem with replacing our read only views with client calculated columns is that some of these calculations use data from related tables. I suspect that it’s a tall order for calculated columns to be able to leverage implicit joins? 😅

**Comment by lukaseder** [comment]
Thanks a lot for your valuable input, as always, @alf. So, how I understand this, you're mostly interested in the `DataType.virtual()` generation option (generate on read), not the `DataType.stored()` one (generate on write). As far as I know, Oracle only supports "virtual" computed columns, though other RDBMS may support both types.

So far, the implementation I've been working on implements the `STORED` semantics, as that seems to be much harder to do, where `INSERT`, `UPDATE`, and `MERGE` statements need to be transformed. Compared to that, the `VIRTUAL` semantics seemingly only needs to replace the column reference by its computation expression.

How this works so far, there's a `<forcedType/>` that specifies a `<generator/>` on a column:

```xml
<forcedType>
    <generator>() -> VAL.times(2)</generator>
    <includeExpression>(?i:T_COMPUTED_CLIENT_SIDE\.CD)</includeExpression>
</forcedType>
```

It must implement the new `org.jooq.Generator` type:

```java
/**
 * A generator can be used with {@link DataType#generatedAlwaysAs(Generator)} to
 * implement dynamic, client side computed columns.
 */
@FunctionalInterface
public interface Generator<T> extends Supplier<Field<T>>, Serializable {}
```

It can either reference an implementation class on the classpath, or contain an "inline implementation", i.e. a Java code snippet that will be added to generated code, e.g.

```java
/**
 * The column <code>PUBLIC.T_COMPUTED_CLIENT_SIDE.CD</code>.
 */
public final TableField<TComputedClientSideRecord, Integer> CD = createField(
    DSL.name("CD"), SQLDataType.INTEGER.nullable(false), this, "", null, null, () -> VAL.times(2));
```

This works well for `STORED` semantics. For `VIRTUAL` semantics, `<forcedType/>` might not work, but a new `<syntheticObjects/>` child element, which declares the generator. But the key feature here is that a `Generator<T>` is a `Supplier<Field<T>>`, which can use all the columns of the relevant table, as well as other SQL expressions. It can obviously also contain values generated in the client, and wrap them using `DSL.val()` or `DSL.inline()`.

From your description, it becomes clear that a stateless `Supplier<Field<T>>` might not be the best design. I've already contemplated a `Function<Table<?>, Field<T>>` to at least give some table context to the generator, in order to be able to re-use generators across tables. Your examples lead me to believe that there is need for even more abstraction, namely:

```java
interface GeneratorContext {
    /** The table on which we're generating data */
    Table<?> table();
    /** The field whose data we're generating */
    TableField<?> field();
    /** The scope giving access to Scope::configuration, Scope::data, Scope::settings, etc */
    Scope scope();
    // ... more context in the future
}
```

With that, it will be possible for you to store values in `scope().data()`, e.g. via `scope().data().computeIfAbsent(...)` in order to make sure that a value is computed only once per query. Not sure if you need that, but it certainly sounds like something a client side computed `VIRTUAL` column might need, especially if the computed value is computed in Java, prior to running the query. As such, `Generator<T>` would look like this:

```java
@FunctionalInterface
public interface Generator<T> extends Function<GeneratorContext, Field<T>>, Serializable {}
```

And your specific example:

```java
dsl.select(T.relay_id)
   .from(T)
   .where(T.relay_id.eq(input))
```

Here, you get to decide whatever `T.RELAY_ID` really means, and whether it's calculated entirely in the client (Java) or if it just maps to some SQL expression that is computed on the server (though the expression would be created dynamically on the client).

> Having this functionality would complete the Trinity and make jOOQ divine 😂

Does the above sound divine enough? :)

> I suspect that it’s a tall order for calculated columns to be able to leverage implicit joins? 😅

Given that your `Generator<T>` can produce any expression, I don't think implicit joins are too much of an ask. There are a few caveats so far, most of which should be addressed:

- I'm not yet sure what to do about aliased tables. The reason why I suggested `Function<Table<?>, Field<T>>` would be to allow for passing the aliased table to your `Generator`, allowing you to use that again e.g. to dereference an implicit join path in a type safe way (the actual `Function` would need more generic type variables to achieve type safety: `Generator<R extends Record, T extends Table<R>, U> extends Function<T, Field<U>>, Serializable`). With the `GeneratorContext`, it wouldn't be type safe, but we could still add the other parameters too for that.
- If the table can't be passed to your `Generator` in a type safe way, then the only solution is to avoid qualification. In the existing implementation for the `STORED` semantics, this didn't seem to cause any trouble yet, though I still have to add more tests for multi-table `UPDATE` statements, and the `MERGE` statement is still quite unexplored. For the `VIRTUAL` semantics, however, removing table qualification doesn't seem to be an option. The risk of creating ambiguities is too high. So, this certainly needs to be addressed
- It's not possible yet to use implicit joins in the `STORED` semantics, as DML statements do not yet support the feature: https://github.com/jOOQ/jOOQ/issues/7508. That's probably not relevant to you

In any case, your feedback was very helpful. I think I have a much clearer picture now of what `VIRTUAL` client side computed columns must be able to do.

### Workaround before this is available

Note, you could perhaps be able to work around this missing functionality yourself already using jOOQ 3.16, in 2 steps?

- You can add "virtual" columns using some code generation extension, either by implementing your own `OracleDatabase` from jOOQ-meta, which adds the column definitions to the code generation database, or by using some regex patching mechanism on generated code. The columns could be marked as `readonly()`, which is a jOOQ 3.16 feature, and which prevents the columns from ever being included in your `INSERT`, `UPDATE`, `MERGE` statements.
- You can then patch all generated SQL using a `VisitListener` where you check whether the visited `QueryPart` is, in fact, your `TableField`. Then you replace that by the desired expression.

It's quite laborious, but it might work and could help you prototype this feature here, checking whether this feature here will work correctly, for you. This *might* even work with implicit join expressions, not sure...

**Comment by alf** [comment]
I believe you're correct that we're after the `DataType.virtual()` generation option for now.

> Here, you get to decide whatever T.RELAY_ID really means, and whether it's calculated entirely in the client (Java) or if it just maps to some SQL expression that is computed on the server (though the expression would be created dynamically on the client).

That sounds perfect. I fear you're giving us a tool so powerful that we can easily create an unholy mess if we're not careful. :smiling_imp:

As promised, here are a few actual examples of how we're using implicit joins to calculate data in the database. It's not very pretty, so we're actively looking for better solutions.

```sql
CREATE OR REPLACE VIEW Utdanningsplan_Emne_Resultat AS
SELECT Institusjonsnr_Eier
     , Fodselsdato
     , Personnr
     , Studieprogramkode
     , Arstall_Start
     , Terminkode_Start
     , Institusjonsnr
     , Emnekode
     , Versjonskode
     , Vektingstall_Redusert AS Vektingstall_Planlagt
     , Vekttypekode_Redusert AS Vekttypekode_Planlagt
     , Vekttypekode_Redusert AS Vekttypekode_Oppnadd
     , CASE WHEN
         EXISTS (
           SELECT 1
             FROM FS.Studentvurdkombprotokoll SVP
             JOIN FS.Vurderingskombinasjon VK
               ON     SVP.Institusjonsnr_Eier = VK.Institusjonsnr_Eier
                  AND SVP.Emnekode            = VK.Emnekode
                  AND SVP.Versjonskode        = VK.Versjonskode
                  AND SVP.Vurdkombkode        = VK.Vurdkombkode
            WHERE VK.Vurdordningkode IS NOT NULL
                AND SVP.Vurdresstatkode      = 'B'
                AND SVP.Institusjonsnr_Eier  = TAB.Institusjonsnr_Eier
                AND SVP.Fodselsdato          = TAB.Fodselsdato
                AND SVP.Personnr             = TAB.Personnr
                AND SVP.Institusjonsnr       = TAB.Institusjonsnr
                AND SVP.Emnekode             = TAB.Emnekode
                AND SVP.Versjonskode         = TAB.Versjonskode
         )
         THEN Vektingstall_Redusert
         ELSE 0
         END                 AS Vektingstall_Oppnadd
  FROM
       FS.StudProgStud_Emne TAB
 WITH CHECK OPTION
;
```

```sql
CREATE OR REPLACE VIEW Utdanningsplan_Eksternresultat_Resultat AS
SELECT TAB.Institusjonsnr_Eier
     , TAB.Fodselsdato
     , TAB.Personnr
     , TAB.Studieprogramkode
     , TAB.Arstall_Start
     , TAB.Terminkode_Start
     , TAB.Institusjonsnr
     , TAB.Lopenr
     , TAB.Reslopenr
     , COALESCE( Vektingstall_Kvalifikasjon
               , RES.Vektingstall )                        Vektingstall_Planlagt
     , CASE
           WHEN Vektingstall_Kvalifikasjon IS NOT NULL THEN
               Vekttypekode_Kvalifikasjon
           ELSE RES.Vekttypekode
       END                                                 Vekttypekode_Planlagt
     , CASE
           WHEN RES.Vurdresstatkode = 'B' THEN
               COALESCE( Vektingstall_Kvalifikasjon
                       , RES.Vektingstall )
           ELSE 0
       END                                                 Vektingstall_Oppnadd
     , CASE
           WHEN Vektingstall_Kvalifikasjon IS NULL THEN
               RES.Vekttypekode
           ELSE Vekttypekode_Kvalifikasjon
       END                                                 Vekttypekode_Oppnadd
  FROM
       FS.StudProgStud_EkstRes                                  TAB
  JOIN FS.Person_Ekstresultat                                   RES
    ON     TAB.Institusjonsnr_Eier  = RES.Institusjonsnr_Eier
       AND TAB.Fodselsdato          = RES.Fodselsdato
       AND TAB.Personnr             = RES.Personnr
       AND TAB.Institusjonsnr       = RES.Institusjonsnr
       AND TAB.Lopenr               = RES.Lopenr
  WITH CHECK OPTION
;
```

Thank you for the suggested work around. I'm intrigued by the thought of creating our own `OracleDatabase`. If I understand correctly this will also allow us to fetch information about foreign keys etc from our own data catalog/configuration and use this during code gen. If we can do this in a forward compatible way with jOOQ then that might actually be perfect for us.

**Comment by lukaseder** [comment]
> I fear you're giving us a tool so powerful that we can easily create an unholy mess if we're not careful. 😈

Yeah, things do get more and more powerful with these SPIs 😅

> As promised, here are a few actual examples of how we're using implicit joins to calculate data in the database. It's not very pretty, so we're actively looking for better solutions

I see. So there are 2 use-cases. You have a `TAB` table, and you want to dynamically project additional columns. 

The second example could really be implemented very nicely using implicit joins (note you'd get left join semantics, not inner join semantics, in case the foreign key from `TAB` to `RES` is optional) The resulting query wouldn't look much different, but it would be much easier to write, just project or filter the column, or don't. So, for example, your `Vekttypekode_Oppnadd` column would have a generator of the form 

```java
() -> when(TAB.Vektingstall_Kvalifikasjon.isNull(), TAB.Person_Ekstresultat().Vekttypekode)
      .else_(TAB.Vekttypekode_Kvalifikasjon)
```

The first example could work even better. Instead of the lengthy `EXISTS()` expression, add this generator (more or less):

```java
() -> when(exists(
            selectOne()
            .from(VK)
            .where(VK.svp().tabKey().eq(TAB.key()) // Assuming embeddable keys
          ), TAB.Vektingstall_Redusert)
      .else_(inline(0))
```

I'm not sure in which direction your constraints go (e.g. `VK` -> `SVP` or the other way round), but the result should be similar either way.

That's really *much* more powerful than anything a server side computed column could do, I like it very much! It's totally in the spirit of SQL as understood by jOOQ!

> Thank you for the suggested work around. I'm intrigued by the thought of creating our own `OracleDatabase`. If I understand correctly this will also allow us to fetch information about foreign keys etc from our own data catalog/configuration and use this during code gen. If we can do this in a forward compatible way with jOOQ then that might actually be perfect for us.

Yes, you can either use the `OracleDatabase`'s existing API, or run your own queries against the Oracle dictionary views, in order to introspect your catalog and make decisions based on that. Just note that jOOQ's runtime will not know the difference between actual columns and columns that you have added synthetically.

Just let me know if you have any specific questions about this approach.

**Comment by alf** [comment]
I can hardly wait!

**Comment by alf** [comment]
One more example from us that I just realized would be made trivial by this extension and that in addition would simplify our view generator immensely.

We have a lot of composite keys that contain sensitive information. This is old legacy stuff that is quite entrenched in the original data model and thus hard to fix.

In our view layer we decided to handle this by joining in a table that contains surrogate keys for the offending columns, and then exposing the surrogate instead of the sensitive key to our jOOQ layer.

Divine indeed.

**Comment by alf** [comment]
I just realized that the jOOQ Code Generator [is extensible](https://www.jooq.org/doc/latest/manual/code-generation/codegen-custom-code/). You didn't mention this option in part 1 of your suggested workaround, so I guess it might not be the best approach here, still it's useful information.

**Comment by lukaseder** [comment]
> In our view layer we decided to handle this by joining in a table that contains surrogate keys for the offending columns, and then exposing the surrogate instead of the sensitive key to our jOOQ layer.

Yeah, I guess the interesting feature here could be to:

- Generate a client side computed column that produces the surrogate key
- Treat the sensitive key as a "client side hidden column": https://github.com/jOOQ/jOOQ/issues/13411, i.e. a column that is only visible to some internal code of yours, including the client side computed expressions.

That way, you won't accidentally project the column (unless you're using plain SQL, of course)

**Comment by alf** [comment]
That would be nice! Right now we actually expose both to our API developers, and trust that they don’t do something stupid. 😅

**Comment by lukaseder** [comment]
Well, you could generate deprecation on the columns, with a custom deprecation remark?
https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-comments/

**Comment by lukaseder** [comment]
So, the configurational difference between `STORED` and `VIRTUAL` computed columns lies in the fact that `STORED` semantics is achieved when the column exists in the actual database, whereas `VIRTUAL` semantics is achieved when the `<forcedType/>` configuration matches a synthetic column (to be implemented: https://github.com/jOOQ/jOOQ/issues/13434). The rest of the configuration should be identical.

**Comment by lukaseder** [comment]
So, I've made some progress today, will check in shortly. The next snapshot build will allow for something like this:

```sql
CREATE TABLE computed_columns_parent (
  parent_id int,
  parent_val int,
  
  CONSTRAINT pk_computed_columns_parent PRIMARY KEY (parent_id)
);

CREATE TABLE computed_columns_client_side_virtual (
  id int,
  parent_id int REFERENCES computed_columns_parent,
  val int,
  
  CONSTRAINT pk_computed_columns_client_side_virtual PRIMARY KEY (id)
);
```

And then:

```xml
<configuration>
    <generator>
        <database>
            <name>org.jooq.meta.extensions.ddl.DDLDatabase</name>
            <properties>
                <property>
                    <key>scripts</key>
                    <value>${basedir}/src/main/resources/computedcolumnsclientsidevirtual.sql</value>
                </property>
            </properties>
            <syntheticObjects>

                <!-- First declare synthetic columns that don't exist in the underlying schema -->
                <columns>
                    <column>
                        <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL</tables>
                        <name>I</name>
                        <type>INTEGER</type>
                    </column>
                    <column>
                        <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL</tables>
                        <name>J</name>
                        <type>INTEGER</type>
                    </column>
                    <column>
                        <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL</tables>
                        <name>K</name>
                        <type>INTEGER</type>
                    </column>
                    <column>
                        <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL</tables>
                        <name>P_ID</name>
                        <type>INTEGER</type>
                    </column>
                    <column>
                        <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL</tables>
                        <name>P_VAL</name>
                        <type>INTEGER</type>
                    </column>
                </columns>
            </syntheticObjects>
            <forcedTypes>

                <!-- Then attach a generator to each of the synthetic columns -->
                <forcedType>
                    <generator>() -> org.jooq.impl.DSL.val(1)</generator>
                    <includeExpression>I</includeExpression>
                </forcedType>
                <forcedType>
                    <generator>() -> org.jooq.impl.DSL.val("2")</generator>
                    <name>VARCHAR(10)</name>
                    <includeExpression>J</includeExpression>
                </forcedType>
                <forcedType>
                    <generator>() -> org.jooq.impl.DSL.val("3")</generator>
                    <userType>java.lang.String</userType>
                    <converter>org.jooq.Converter.ofNullable(Integer.class, String.class, Object::toString, Integer::valueOf)</converter>
                    <includeExpression>K</includeExpression>
                </forcedType>
                <forcedType>
                    <generator>() -> computedColumnsParent().PARENT_ID</generator>
                    <includeExpression>P_ID</includeExpression>
                </forcedType>
                <forcedType>
                    <generator>() -> computedColumnsParent().PARENT_VAL</generator>
                    <includeExpression>P_VAL</includeExpression>
                </forcedType>
            </forcedTypes>
        </database>
        <generate>
            <implicitJoinPathsToOne>true</implicitJoinPathsToOne>
            <records>true</records>
        </generate>
        <target>
            <packageName>org.jooq.codegen.test.computedcolumnsclientsidevirtual.db</packageName>
        </target>
    </generator>
</configuration>
```

And finally:

```java
@Test
public void testDML() {
    ComputedColumnsClientSideVirtual t = COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL;

    ComputedColumnsParentRecord p1 = ctx.newRecord(COMPUTED_COLUMNS_PARENT);
    p1.setParentId(2);
    p1.setParentVal(3);
    assertEquals(1, p1.store());

    assertEquals(1,
    ctx.insertInto(t)
       .columns(t.ID, t.PARENT_ID)
       .values(1, 2)
       .execute());

    Record r1 = ctx.fetchSingle(t);
    assertEquals(1, r1.get(t.ID));
    assertNull(r1.get(t.VAL));
    assertEquals(1, r1.get(t.I));
    assertEquals("2", r1.get(t.J));
    assertEquals("3", r1.get(t.K));
    assertEquals(2, r1.get(t.P_ID));
    assertEquals(3, r1.get(t.P_VAL));

    assertEquals(1, ctx.update(t).set(t.VAL, 1).execute());

    Record r2 = ctx.fetchSingle(t);
    assertEquals(1, r2.get(t.ID));
    assertEquals(1, r2.get(t.VAL));
    assertEquals(1, r2.get(t.I));
    assertEquals("2", r2.get(t.J));
    assertEquals("3", r2.get(t.K));
    assertEquals(2, r2.get(t.P_ID));
    assertEquals(3, r2.get(t.P_VAL));

    println(ctx.select(t.ID).from(t));
    println(ctx.selectFrom(t));
}
```

The two `println` actually print these queries here:

```sql
-- First println
select "COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL"."ID"
from "COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL"

-- Second println
select
  "COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL"."ID",
  "COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL"."PARENT_ID",
  "COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL"."VAL",
  1,
  '2',
  '3',
  "alias_27107310"."PARENT_ID",
  "alias_27107310"."PARENT_VAL"
from (
  "COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL"
    left outer join "COMPUTED_COLUMNS_PARENT" "alias_27107310"
      on "COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL"."PARENT_ID" = "alias_27107310"."PARENT_ID"
  )
```

As expected the mere inclusion of the computed columns triggers the implicit join logic, or whatever the computation implies.

There are still lots of rough spots, such as:

- being able to also use nested records and multisets in such computed columns
- making sure the computed columns are read only
- the thing about passing the table to the lambda, rather than having a "static" `Supplier<Field<?>>`
- much more

but I'm positive that I can sort them all out. This is truly exciting!

**Comment by alf** [comment]
Nice! I’ll give the nightly a go during Easter.

**Comment by lukaseder** [comment]
I've pushed a new set of changes which improves quite a few things of this feature. The `DEFAULT` / `GENERATED BY DEFAULT AS` semantics will be postponed, as it is less interesting compared to the `GENERATED ALWAYS AS` semantics: https://github.com/jOOQ/jOOQ/issues/13443

**Comment by lukaseder** [comment]
A test I've just added makes some transitive computations. Actual tables:

```sql
CREATE TABLE computed_columns_parent (
  parent_id int,
  parent_val int,
  
  CONSTRAINT pk_computed_columns_parent PRIMARY KEY (parent_id)
);

CREATE TABLE computed_columns_client_side_virtual (
  id int,
  parent_id int REFERENCES computed_columns_parent,
  val int,
  
  CONSTRAINT pk_computed_columns_client_side_virtual PRIMARY KEY (id)
);

CREATE TABLE computed_columns_client_side_virtual_transitive (
  id int,
  parent_id int REFERENCES computed_columns_client_side_virtual,
  val int,
  
  CONSTRAINT pk_computed_columns_client_side_virtual_transitive PRIMARY KEY (id)
);
```

Computed columns:

```xml
<syntheticObjects>
    <columns>
        <column>
            <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL</tables>
            <name>I</name>
            <type>INTEGER</type>
        </column>
        <column>
            <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL</tables>
            <name>J</name>
            <type>INTEGER</type>
        </column>
        <column>
            <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL</tables>
            <name>K</name>
            <type>INTEGER</type>
        </column>
        <column>
            <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL</tables>
            <name>VAL2</name>
            <type>INTEGER</type>
        </column>
        <column>
            <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL</tables>
            <name>P_ID</name>
            <type>INTEGER</type>
        </column>
        <column>
            <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL</tables>
            <name>P_VAL</name>
            <type>INTEGER</type>
        </column>
        
        <column>
            <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL_TRANSITIVE</tables>
            <name>VAL_COMPUTED1</name>
            <type>INTEGER</type>
        </column>
        <column>
            <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL_TRANSITIVE</tables>
            <name>VAL_COMPUTED_COMPUTED</name>
            <type>INTEGER</type>
        </column>
        <column>
            <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL_TRANSITIVE</tables>
            <name>VAL_COMPUTED2</name>
            <type>INTEGER</type>
        </column>
        <column>
            <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL_TRANSITIVE</tables>
            <name>P_VAL</name>
            <type>INTEGER</type>
        </column>
    </columns>
</syntheticObjects>
<forcedTypes>
    <forcedType>
        <generator>ctx -> org.jooq.impl.DSL.val(1)</generator>
        <includeExpression>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL\.I</includeExpression>
    </forcedType>
    <forcedType>
        <generator>ctx -> org.jooq.impl.DSL.val("2")</generator>
        <name>VARCHAR(10)</name>
        <includeExpression>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL\.J</includeExpression>
    </forcedType>
    <forcedType>
        <generator>ctx -> org.jooq.impl.DSL.val("3")</generator>
        <userType>java.lang.String</userType>
        <converter>org.jooq.Converter.ofNullable(Integer.class, String.class, Object::toString, Integer::valueOf)</converter>
        <includeExpression>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL\.K</includeExpression>
    </forcedType>
    <forcedType>
        <generator>ctx -> VAL.times(2)</generator>
        <includeExpression>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL\.VAL2</includeExpression>
    </forcedType>
    <forcedType>
        <generator>ctx -> computedColumnsParent().PARENT_ID</generator>
        <includeExpression>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL\.P_ID</includeExpression>
    </forcedType>
    <forcedType>
        <generator>ctx -> computedColumnsParent().PARENT_VAL</generator>
        <includeExpression>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL\.P_VAL</includeExpression>
    </forcedType>

    <forcedType>
        <generator>ctx -> VAL.plus(1)</generator>
        <includeExpression>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL_TRANSITIVE\.VAL_COMPUTED1</includeExpression>
    </forcedType>
    <forcedType>
        <generator>ctx -> VAL.plus(2)</generator>
        <includeExpression>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL_TRANSITIVE\.VAL_COMPUTED2</includeExpression>
    </forcedType>
    <!-- Lambda isn't possible here, because of forward reference -->
    <forcedType>
        <generator><![CDATA[
            new org.jooq.Generator<Integer>() {
                @Override
                public Field<Integer> apply(org.jooq.GeneratorContext ctx) {
                    return VAL_COMPUTED1.plus(VAL_COMPUTED2);
                }
            }
        ]]></generator>
        <includeExpression>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL_TRANSITIVE\.VAL_COMPUTED_COMPUTED</includeExpression>
    </forcedType>
    <forcedType>
        <generator>ctx -> computedColumnsClientSideVirtual().P_VAL</generator>
        <includeExpression>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL_TRANSITIVE\.P_VAL</includeExpression>
    </forcedType>
</forcedTypes>
```

Test case:

```java
    @Test
    public void testTransitiveComputedColumns() {
        insertData();

        ComputedColumnsClientSideVirtualTransitive t = COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL_TRANSITIVE;
        ComputedColumnsClientSideVirtualTransitiveRecord r = ctx.fetchSingle(t);

        println(ctx.selectFrom(t.as("t")));

        assertEquals(3, r.getValComputed1());
        assertEquals(4, r.getValComputed2());
        assertEquals(7, r.getValComputedComputed());
        assertEquals(3, r.getPVal());
    }
```

The test case passes with the inserted data. The computations can be seen here from the println output:

```sql
select
  "t"."ID",
  "t"."PARENT_ID",
  "t"."VAL",

  -- Transitive expressions on the same table
  ("t"."VAL" + 1) "VAL_COMPUTED1",
  (("t"."VAL" + 1) + ("t"."VAL" + 2)) "VAL_COMPUTED_COMPUTED",
  ("t"."VAL" + 2) "VAL_COMPUTED2",

  -- Transitive implicit join
  "alias_43499055"."PARENT_VAL" "P_VAL"
from (

  -- Aliasing of table is possible, too
  "COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL_TRANSITIVE" "t"
    left outer join (
      "COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL" "alias_71513992"
        left outer join "COMPUTED_COLUMNS_PARENT" "alias_43499055"
          on "alias_71513992"."PARENT_ID" = "alias_43499055"."PARENT_ID"
    )
      on "t"."PARENT_ID" = "alias_71513992"."ID"
  )
```

Other integration tests check if computed columns can be correctly used in expressions, such as the above addition, function calls, aggregate functions, window functions, in the `WHERE` clause and elsewhere. It seems that the implementation is quite sound. I'll test for a few more edge cases, but I think this can ship as a non-experimental feature after all.

**Comment by alf** [comment]
Please forgive me for asking, but do you have a due date for 3.17? This is insane.

**Comment by lukaseder** [comment]
Roughly end of Q2 2022, i.e. in ~ 2-3 months? There had been a few big projects, which all need to stabilise. Especially the generated multiset convenience is far from ready yet: https://github.com/jOOQ/jOOQ/issues/13069

And while `VIRTUAL` client side computed columns seem sound enough, the `STORED` ones still have a lot of open questions / feature interactions, which haven't been tested yet.

Also, with all these computed columns, I wanted to tackle a few other power user features, finally, including:

- Audit columns: https://github.com/jOOQ/jOOQ/issues/1592
- Optimistic locking also for statements: https://github.com/jOOQ/jOOQ/issues/13339

Though those last two could also be postponed... Though, I like their "dogfooding" nature to make sure the computed columns respond even better to real world use-cases.

**Comment by alf** [comment]
That sounds great. Thank you.

**Comment by lukaseder** [comment]
It's just a distraction from a test case. The columns are added in this order:

```xml
        <column>
            <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL_TRANSITIVE</tables>
            <name>VAL_COMPUTED1</name>
            <type>INTEGER</type>
        </column>
        <column>
            <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL_TRANSITIVE</tables>
            <name>VAL_COMPUTED_COMPUTED</name>
            <type>INTEGER</type>
        </column>
        <column>
            <tables>COMPUTED_COLUMNS_CLIENT_SIDE_VIRTUAL_TRANSITIVE</tables>
            <name>VAL_COMPUTED2</name>
            <type>INTEGER</type>
        </column>
```

Meaning the computed column in the middle depends on a column that was placed before and one that was placed after. The test needs to make sure that this computation as well as the class and instance initialisation still works, as this order should be irrelevant. However, the Java compiler didn't let me refer to a field that is located after the lambda, which in turn is part of the instance initialiser.

A simpler example:

```java
class X {
  int i;
  Runnable r = () -> System.out.println(i + j);
  int j;
}
```

The compiler output is:

```
jshell> class X {
   ...>     int i;
   ...>     Runnable r = () -> System.out.println(i + j);
   ...>     int j;
   ...> }
|  Error:
|  illegal forward reference
|      Runnable r = () -> System.out.println(i + j);
|                                                ^
```

So, accessing `i` is fine, but accessing `j` isn't. This, however, is fine:

```java
class X {
  int i;
  Runnable r = new Runnable() { public void run() { System.out.println(i + j); }};
  int j;
}
```

This might be sorted if the `GeneratorContext` is changed to `GeneratorContext<R extends Record, T extends Table<R>>` where `T` is bound to the table type, and as such, `ctx.table().VAL_COMPUTED` could offer type safe access to table columns without the lambda needing to capture `this`.

Not sure if this makes sense? It's just an edge case. Users will define the columns in a more "incremental" order.

**Comment by alf** [comment]
Thank you for the explanation. That cleared things up for me.

**Comment by lukaseder** [comment]
Interestingly, this works, too:

```java
class X {
  int i;
  Runnable r = () -> System.out.println(i + this.j); // "this." prefix
  int j;
}
```

See: https://bugs.openjdk.java.net/browse/JDK-8027941

**Comment by lukaseder** [comment]
As far as I can tell, this is quite complete for most dialects, with some caveats in some dialects that may have limitations in `MERGE`, etc. A follow-up task to address known issues has been created: https://github.com/jOOQ/jOOQ/issues/13685

**Comment by lukaseder** [comment]
Forgot to integration test `KotlinGenerator` and `ScalaGenerator`. Examples don't compile yet...
